### ISSUE TITLE
Feature request: remove document by id

### ISSUE THREAD

**Issue by kdekooter** [lucaong/minisearch#22]

**Comment by lucaong** [comment]
Hi @kdekooter, thanks for your feature request!
I know that removing a document by id would be convenient, but unfortunately in order to do so, MiniSearch would have to save all fields that were indexed, in order to be able to remove them from the inverted index. This would mean that MiniSearch would have to "hold" on the original documents, not only on the index. This might not be what the user wants, as someone that is only interested in searching through specific fields in a large collection of documents, but not in returning their value, would significantly reduce the memory utilization by not keeping a reference to the original documents.

As it is reasonably easy for a user to keep a map of document ID to document, MiniSearch leaves the choice to the user. Here is an example of how that can be done efficiently:

```javascript
// Build a map of id to document
const documentById = documents.reduce(function (docById, document) {
  docById[document.id] = document
  return docById
},{})

// Later, you can use the map to
// delete a document:
miniSearch.remove(documentById[id])
```

That said, this feature could be offered by higher-level libraries built on top of MiniSearch (many of them would anyway keep a map of id to document internally). For example, after your request, I will soon add this to the newly-released React integration: https://github.com/lucaong/react-minisearch 

I don't know if you use React in your project, but if so you might find the library useful. Otherwise, you could take inspiration for your integration.

**Comment by kdekooter** [comment]
Thank you for your elaborate response @lucaong! My project is built with VueJS (you might want to check it out: https://notestack.io) so unfortunately I can't use your React component. 

Adding a map storing the original document only cost me 3 lines of code though ;-).

**Comment by lucaong** [comment]
Hi @kdekooter , thanks for the feedback.

Your project looks amazing! I am happy that you are considering MiniSearch for it, and I am sure you won't be disappointed :)

I am not very fluent with VueJS, but if I find the time I would like to work on a Vue integration, to make it easier for Vue users to adopt MiniSearch.

**Comment by kdekooter** [comment]
Thanks for the compliment!

I quickly wrote up a short blogpost demonstrating how I use MiniSearch at the moment: https://notestack.io/public/fulltext-search-your-vuex-store-with-minisearch/aa0e7b05-f5d6-43d7-8206-4eafdaac94e6.

**Comment by lucaong** [comment]
@kdekooter I revive this old thread to announce that the newly released v6.0.0-beta.1 (soon to be v6.0.0) includes a new method `discard` (and related methods `discardAll` and `replace`) which removes a document by ID.

The reasons given in this thread for why remove takes a full document are still valid, but discard takes a different approach, made possible by the internal design of MiniSearch. Instead of immediately removing a document from the index (which would only be feasible given the original document), discard marks the current document version as discarded, so it is immediately ignored by searches, but does not mutate the index. Instead, that is left to a vacuuming process, ran by default after a certain number of documents are discarded. Vacuuming traverses the index and removes references to discarded documents, allowing memory to be released. See the documentation here: https://lucaong.github.io/minisearch/classes/_minisearch_.minisearch.html#discard

This provides the feature requested by this issue, while avoiding the drawbacks of a naive implementation: the original documents are not stored, and use cases that do not need to remove documents incur no overhead.
